File: tia-client/src/main/java/com/teamscale/tia.client/UrlUtils.java
package com.teamscale.tia.client;

public class UrlUtils {

	/**
	 * Percent-decodes a string, such as used in a URL Path (not a query string / form encode, which uses + for spaces,
	 * etc)
	 */
	public static String percentEncode(String encodeMe) {
		if (encodeMe == null) {
			return "";
		}
		String encoded = encodeMe.replace("%", "%25");
		encoded = encoded.replace(" ", "%20");
		encoded = encoded.replace("!", "%21");
		encoded = encoded.replace("#", "%23");
		encoded = encoded.replace("$", "%24");
		encoded = encoded.replace("&", "%26");
		encoded = encoded.replace("'", "%27");
		encoded = encoded.replace("(", "%28");
		encoded = encoded.replace(")", "%29");
		encoded = encoded.replace("*", "%2A");
		encoded = encoded.replace("+", "%2B");
		encoded = encoded.replace(",", "%2C");
		encoded = encoded.replace("/", "%2F");
		encoded = encoded.replace(":", "%3A");
		encoded = encoded.replace(";", "%3B");
		encoded = encoded.replace("=", "%3D");
		encoded = encoded.replace("?", "%3F");
		encoded = encoded.replace("@", "%40");
		encoded = encoded.replace("[", "%5B");
		encoded = encoded.replace("]", "%5D");
		return encoded;
	}
}

File: tia-client/src/main/kotlin/com/teamscale/tia/client/UrlUtils.kt
package com.teamscale.tia.client

/** Utils for URL encoding as the Java internal URLEncoder does not handle all escape-worthy symbols for path segments.  */
object UrlUtils {
	/**
	 * Percent-decodes a string, such as used in a URL Path (not a query string / form encode, which uses + for spaces,
	 * etc).
	 * Source: https://stackoverflow.com/a/44076794
	 */
	@JvmStatic
	fun String.encodeUrl() =
		replace("%", "%25")
			.replace(" ", "%20")
			.replace("!", "%21")
			.replace("#", "%23")
			.replace("$", "%24")
			.replace("&", "%26")
			.replace("'", "%27")
			.replace("(", "%28")
			.replace(")", "%29")
			.replace("*", "%2A")
			.replace("+", "%2B")
			.replace(",", "%2C")
			.replace("/", "%2F")
			.replace(":", "%3A")
			.replace(";", "%3B")
			.replace("=", "%3D")
			.replace("?", "%3F")
			.replace("@", "%40")
			.replace("[", "%5B")
			.replace("]", "%5D")
}
